### ISSUE TITLE
Need a data sampler for unevenly distributed labels

### ISSUE THREAD

**Issue by apple2373** [chainer/chainer#2860]
We need a way to handle a dataset whose label distribution is highly skewed. For example, when we have 1000 positives and 100 negatives, we want to make sure each batch contains the same number of positives and negatives, oversampling negative examples 10 times more than positives. 

Someone said pytorch has a way to do:
Adding a `sampler` option for iterator as PyTorch can solve the problem I guess.
http://pytorch.org/docs/data.html#torch.utils.data.DataLoader
Source: https://chainer.slack.com/archives/C0LC5A6C9/p1497343348496751
Maybe it’s possible for chainer to have a wrapper for these pytoarch data preprocessing utilities. DataLorder is not involved in gradient computation so it should be easy and take much less time than implementing equivalent functions from scratch.

**Comment by hvy** [comment]
Not directly addressing the same issue, but related. https://github.com/chainer/chainer/pull/3429

**Comment by hvy** [comment]
Bump.

**Comment by beam2d** [comment]
#3429 looks related.

**Comment by apple2373** [comment]
Thank you @mitmul ! Just to clarify I already have my own (dirty coded) data loader so i am not looking for the solution now. But yes, the BalancedSerialIterator is what I was looking for when I originally created the issue. I wonder if it's a best way to keep it only for Chainer Chemistry. I think class imbalance problem appears everywhere if you are working on real world problems... 

PR: #3429  is also related but not exactly what I was looking for because I still need to code sampler by myself. 

Probably the best solution is to provide both sampler option (i.e., PR: #3429) and then also provide some default sampler for class imbalance problem (i.e., a sampler equivalent to BalancedSerialIterator) ?

**Comment by corochann** [comment]
Thank you @apple2373 for the feedback.
I implemented `BalancedSerialIterator`, and I agree with your opinion.

I commented https://github.com/chainer/chainer/pull/3429

> After this PR merged, I think it is nice to introduce BalancedOrderSampler rather than BalancedSerialIterator (used in Chainer Chemistry until now) so that the sampler can be reused among any kinds of Iterators.

**Comment by hvy** [comment]
Fyi, #3429 is now merged. It would be great if we could provide some sort of balanced sampler.
